src/riscv/lib/src/machine_state/page_cache/state.rs

```diff
@@ -278,7 +284,25 @@ impl<CPE: CodePageEntry<MC, M>, MC: MemoryConfig, M: ManagerBase> PageCache<MC,
             return;
         }
 
-        *page_entry = PageEntry::new::<PopulationError>(self.compiler.clone(), |offset| {
+        let fetch_compiler = || {
+            let idx = page_index as u64;
+            if let Some(compiler) = self.router.get(&idx) {
+                compiler.clone()
+            } else {
+                #[cfg(feature = "log")]
+                crate::log::warning!(
+                    "Populating page requires ad-hoc single page range to be added in router: {page_index}"
+                );
+                self.router
+                    .add_range(idx..=idx, || CPE::new_compiler(&self.compiler_context));
```

Comment:
I wonder if `add_range` could return an `&T` of the value that has just been inserted?